### DISCUSSION TITLE
I'm looking for a way to validate something like: "a where clause on the status column must be included" before Slick executes the SQL

### DISCUSSION THREAD

**Discussion by tanin47** [slick/slick#3093] [Questions]
Is there a way to hook into some stages of Slick?

If there's a way that I can insert my logic after SQL is generated, that would work. I can parse the SQL, validate, and decide to throw an exception or not.

**Comment by nafg** [comment]
I'm not sure offhand. A lot of things can be overridden but I'm not sure if this is one of them.

Can you elaborate on your use case?

**Reply by tanin47** [reply]
To simplify the situation, we have a database table: `post` that contains the `status` column. The `status` can be `Active` and `Deleted`. Now, in other places, we keep forgetting to filter by `status` when we query. It has caused bugs here and there.

Since I'm really big of code safety, I want to add a validation logic in some layers in order to enforce that `status` must exist in the `WHERE` clause. I would expect this validation would almost always fail in tests if our team member forget.

**Reply by nafg** [reply]
Is this only for selects or also inserts/updats/deletes?

You could use views like `create view active_posts as select * from post where status='Active'` etc.

**Reply by nafg** [reply]
What about just hiding your table behind a service? Make it private and the code can only interact with it through a service class that implements these rules once and for all.

**Reply by tanin47** [reply]
Only SELECT for now.

> You could use views like create view active_posts as select * from post where status='Active' etc.

A view could work but we would have to have 2 queries: one for SELECT a list of things, and another for SELECT on a specific `id` which wouldn't the `status` column. It's brittle when the validation gets more complex.

Or does Slick support switching the table name on the fly (i.e. without defining a new table)?

**Reply by nafg** [reply]
Well you could use normal code constructs, like

```scala
class PostsLike(tag: Tag, tableName: String) extends Table(tag, tableName) {
  val id = column....
  ...

  val * = ....
}

class Posts(tag: Tag) extends PostsLike(tag, "post")
val Posts = TableQuery[Posts]

class ActivePosts(tag: Tag) extends PostsLike("tag", "active_posts")
val ActivePosts = TableQuery[ActivePosts]

...
```

**Reply by nafg** [reply]
I can also chat in realtime on Discord if you like

**Reply by tanin47** [reply]
Thank you. I think it's decent. The only con I can see is that it doesn't enforce people to use `ActivePosts`. probably acceptable to some degree.

Is there a way to hook or override `TableQuery` (maybe with my own class or implicit conversion) where I can add a check where `.result` in invoked? (not sure if I can actually read the query structure)

**Reply by nafg** [reply]
`.result` is provided through an implicit conversion, so you could interfere with it to prevent it in different ways.

But again, I think interface hiding is a better solution to this kind of problem.

**Reply by tanin47** [reply]
Thank you. I'll explore a bit but yeah interface hiding seems like a decent solution.
